WordPress/wordpress-develop
# Section Styles: improve performance and conceptual consistency

Trac ticket https://core.trac.wordpress.org/ticket/61451
Follow-up work for https://core.trac.wordpress.org/ticket/61312

Backports https://github.com/WordPress/gutenberg/pull/62712

The changes in https://github.com/WordPress/gutenberg/pull/62712 make optimizations around the processing and registration of shared block style variations in theme.json as well as simplify the mental model for the feature increasing conceptual consistency.

These changes involve:
- Moving shared variation definitions from `styles.blocks.variations` to `styles.variations`
- Removing `blockTypes` from `styles.variations`. 
- Not automatically registering shared variations from theme style variation or primary theme.json files.
- Moving the merging of theme.json data into the `WP_Theme_JSON_Resolver` and `WP_Theme_JSON` classes as per [#6868](https://github.com/WordPress/wordpress-develop/pull/6868)


## Testing Instructions

1. Defines block style variations via the different sources below:
   - `register_block_style`
   - Theme.json partial file under `/styles` directory including `blockTypes`
   - Within a theme style variation under `styles.variations` 
       - Define at least one that is registered via `register_block_style` or a theme.json partial as well
       - Define a variation that hasn't been registered
2. Confirm that only registered variations appear for selection (those defined via `register_block_style`, theme.json partials, or in core block types `styles` property such as Button's Outline style)
3. Try out all the variations and ensure they get applied correctly in the editor and frontend
4. Update the variation definitions to define inner element and block type styles as well
5. Check the inner element and block type styles are applied correctly for variations
4. Test applying variations on blocks in a nested fashion
5. Update variation styles via Styles > Blocks > Block Type > Variation and confirm they apply correctly everywhere

<details>
<summary><strong>`register_block_style` Example</strong></summary>

```php
gutenberg_register_block_style(
	array( 'core/group', 'core/columns' ),
	array(
		'name'       => 'section-b',
		'label'      => __( 'Section B' ),
		'style_data' => array(
			'color'    => array(
				'background' => '#4f6f52',
				'text'       => '#d2e3c8',
			),
			'blocks'   => array(
				'core/group' => array(
					'color' => array(
						'background' => '#739072',
						'text'       => '#e3eedd',
					),
				),
			),
			'elements' => array(
				'link' => array(
					'color'  => array(
						'text' => '#ead196',
					),
					':hover' => array(
						'color' => array(
							'text' => '#ebd9b4',
						),
					),
				),
			),
		),
	)
);

```
</details>

<details>
<summary><strong>Theme.json Partial Example</strong></summary>

```json
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 3,
	"title": "Section A",
	"slug": "section-a",
	"blockTypes": [ "core/group" ],
	"styles": {
		"color": {
			"background": "slategrey",
			"text": "snow"
		},
		"blocks": {
			"core/group": {
				"color": {
					"background": "darkslategrey",
					"text": "whitesmoke"
				}
			}
		}
	}
}
```
</details>


<details>
<summary><strong>Theme Style Variation Example (`styles.variations`)</strong></summary>

```json
{

	"styles": {
		"variations": {
			"section-a": {
				"color": {
					"background": "var(--wp--preset--color--theme-2)",
					"text": "var(--wp--preset--color--theme-5)"
				},
				"blocks": {
					"core/group": {
						"color": {
							"background": "var(--wp--preset--color--theme-4)",
							"text": "var(--wp--preset--color--theme-2)"
						}
					}
				}
			}
		}
	}
}
```
</details>
